### PR TITLE
maven build

### PR DESCRIPTION
**Pull request by cseed**
@jigold, I made the following modifications to your changes:
 - added a jars target to the hail/ Makefile, which does the C build and then runs maven, and removed the corresponding section of the pom.xml,
 - run the Java tests in its own build step invoking java directory, so I removed the test step from pom.xml
 - modified hail_build_image (the image used to run the Hail build) to install maven and the Hail dependencies (jars), no longer using gradle

The testng tests probably need better logging which I'm trying to figure out now.

### CODE REVIEW

File: hail/Makefile [hunk 1]
```diff
@@ -1,4 +1,12 @@
-.PHONY: shadowJar
+.PHONY: shadowJar jars clean
 
 shadowJar:
 	./gradlew shadowJar
+
+jars:
+	make -C src/main/c
+	mvn compile test-compile jar:jar jar:test-jar assembly:single
```
Comment (jigold) [inline hail/Makefile]: Doesn't assembly:single build both the jar and test-jar already? Why do we need the two jar steps?

File: apiserver/Dockerfile.hail-base [hunk 2]
```diff
@@ -6,8 +6,8 @@ RUN unzip -q hail.zip && \
   mkdir -p /hail/python && \
   mv inner /hail/python/hail
 
-COPY hail-all-spark.jar /hail/jars/hail-all-spark.jar
-RUN cp /hail/jars/hail-all-spark.jar $SPARK_HOME/jars/hail-all-spark.jar
+COPY hail.jar /hail/jars/hail.jar
```
Comment (jigold) [inline apiserver/Dockerfile.hail-base]: Does the renaming of the jar affect cloud tools or the deployment steps at all?

Comment (cseed) [reply apiserver/Dockerfile.hail-base]: Yes.  I renamed things throughout to be hail.jar.

File: hail/Makefile (hunk 1 above)
Comment (cseed) [reply hail/Makefile]: No, I found if I didn't do that then the test jar with dependencies didn't have any of the Hail test classes in it.  Of course, I really want to be able to say `mvn assembly:single` only and have it build the minimum dependencies, but I don't know how to do that without more research.

**Comment by jigold** [comment]
You have an error based on where the maven code is being run (not the directory with the pom file).
```
[[1;31mERROR[m] Failed to execute goal [32morg.apache.maven.plugins:maven-dependency-plugin:2.8:resolve[m [1m(default-cli)[m: [1;31mGoal requires a project to execute but there is no POM in this directory (/). Please verify you invoked Maven from the correct directory.[m -> [1m[Help 1][m
[[1;31mERROR[m] 
[[1;31mERROR[m] To see the full stack trace of the errors, re-run Maven with the [1m-e[m switch.
[[1;31mERROR[m] Re-run Maven using the [1m-X[m switch to enable full debug logging.
[[1;31mERROR[m] 
[[1;31mERROR[m] For more information about the errors and possible solutions, please read the following articles:
[[1;31mERROR[m] [1m[Help 1][m http://cwiki.apache.org/confluence/display/MAVEN/MissingProjectException
The command '/bin/sh -c mvn dependency:resolve &&   rm pom.xml test-jar-with-dependencies.xml' returned a non-zero code: 1
```

File: hail/src/main/scala/is/hail/expr/ir/FoldConstants.scala [hunk 3]
```diff
@@ -5,35 +5,18 @@ import is.hail.utils.HailException
...
-              case _ => false
-            } && (canGenerateLiterals || CanEmit(ir.typ))) {
-              Some(try {
-                Literal.coerce(ir.typ, Interpret(ir, optimize = false))
-              } catch {
-                case e: HailException => Die(e.getMessage + e.getStackTrace.map(s => s"\n    $s").mkString(""), ir.typ)
-              })
-            }
-            else None
+      case ir: IR if !IsConstant(ir) &&
```
Comment (cseed) [inline hail/src/main/scala/is/hail/expr/ir/FoldConstants.scala]: @tpoterba are you OK with this?  FoldConstants was an insane mess and I cleaned it up.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/FoldConstants.scala]: won't this run `Die(Str(...))` on every pass of the optimizer? That seems kind of bad. Should we not wrap this in a die node?

**Comment by cseed** [comment]
Just to clarify, I was seeing an infinite loop in Interpret for a 0-ary function via calls to Interpret => Compile => Optimize => FoldConstants => Interpret.

**Comment by tpoterba** [comment]
ack, damn. That's super annoying.

**Comment by tpoterba** [comment]
interpret calling compile calling optimize is a huge bug.

File: hail/src/main/scala/is/hail/expr/ir/FoldConstants.scala (hunk 3 above)
Comment (cseed) [reply hail/src/main/scala/is/hail/expr/ir/FoldConstants.scala]: Sorry, I put back the set of nodes we shouldn't try to fold (including Die).

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/FoldConstants.scala]: I think your solution was simpler, we just need to think through the edge cases and make sure they're handled in the base case.

File: hail/src/test/scala/is/hail/stats/LogisticRegressionModelSuite.scala [hunk 4]
```diff
@@ -140,110 +140,6 @@ class LogisticRegressionModelSuite extends SparkSuite {
     assert(D_==(scoreStats.p, 0.3724319159, tolerance = 1.0E-5))
   }
 
-  @Test def covariatesVsInterceptOnlyR() {
```
Comment (tpoterba) [inline hail/src/test/scala/is/hail/stats/LogisticRegressionModelSuite.scala]: can you make an issue to add this test back in Python / Scala without R dependence and randomly assign? I concluded a few weeks ago that this is an important unit test

File: build.yaml [hunk 5]
```diff
@@ -146,6 +152,33 @@ steps:
...
+      to: /io/resources.tar.gz
+    - from: /hail-test.jar
+      to: /io/hail-test.jar
+    - from: /testng.xml
+      to: /io/testng.xml
+    - from: /testng-cpp-codegen.xml
+      to: /io/testng-cpp-codegen.xml
+   outputs:
+    - from: /io/test-output
+      to: /test-output
```
Comment (jigold) [inline build.yaml]: Does this go to the gas bucket by default? or is this on the local computer?

File: hail/src/test/scala/is/hail/TestUtilsSuite.scala [hunk 6]
```diff
@@ -10,58 +10,6 @@ import org.testng.annotations.{DataProvider, Test}
...
-    }
-    classes.result()
-  }
-
-  def findClasses(dir: File, packageName: String): Array[Class[_]] = {
-    val classes = new ArrayBuilder[Class[_]]()
-    if (!dir.exists())
-      return classes.result()
-
-    val files = dir.listFiles()
```
Comment (jigold) [inline hail/src/test/scala/is/hail/TestUtilsSuite.scala]: Is this not needed anymore? I had put it in because the tests with data providers wouldn't fail if the data provider failed.

File: hail/src/test/scala/is/hail/expr/ir/FoldConstantsSuite.scala [hunk 7]
```diff
@@ -6,19 +6,6 @@ import org.scalatest.testng.TestNGSuite
...
-  @Test def test_let_evaluation() {
-    val ir = Let(
-      "x",
-      I32(5),
-      Let(
-        "y",
-        ApplyBinaryPrimOp(Add(), Ref("x", TInt32()), I32(10)),
-        ApplyBinaryPrimOp(Add(), Ref("y", TInt32()), I32(15))
-      )
-    )
```
Comment (jigold) [inline hail/src/test/scala/is/hail/expr/ir/FoldConstantsSuite.scala]: Why did you delete this test?

File: hail/src/test/scala/is/hail/stats/LogisticRegressionModelSuite.scala (hunk 4 above)
Comment (jigold) [reply hail/src/test/scala/is/hail/stats/LogisticRegressionModelSuite.scala]: I agree. @cseed Can you make an issue?

File: hail/testng-cpp-codegen.xml [hunk 8]
```diff
@@ -0,0 +1,11 @@
+<suite name="CppCodegenSuite" verbose="1">
+    <test name="TestCppCodegen">
+      <classes>
+           <class name="is.hail.nativecode.NativeCodeSuite"/>
+           <method name="is.hail.annotations.AnnotationsSuite.testReadWrite"/>
```
Comment (jigold) [inline hail/testng-cpp-codegen.xml]: I don't see where the environment variable to enable cpp codegen is set. I thought I had added that to pom.xml

File: hail/src/test/scala/is/hail/stats/LogisticRegressionModelSuite.scala (hunk 4 above)
Comment (cseed) [reply hail/src/test/scala/is/hail/stats/LogisticRegressionModelSuite.scala]: Done, thanks: https://github.com/hail-is/hail/issues/6080

File: build.yaml (hunk 5 above)
Comment (cseed) [reply build.yaml]: These get copied from the local computer (in /io/test-output) into the CI service account bucket with the other artifacts for this build.

File: hail/src/test/scala/is/hail/TestUtilsSuite.scala (hunk 6 above)
Comment (cseed) [reply hail/src/test/scala/is/hail/TestUtilsSuite.scala]: No, it's broken, I created another issue: https://github.com/hail-is/hail/issues/6079.  The code didn't work for a test jar, but skipped test show up as failures so our test coverage isn't diminished (but readable error messages might be).

File: hail/src/test/scala/is/hail/expr/ir/FoldConstantsSuite.scala (hunk 7 above)
Comment (cseed) [reply hail/src/test/scala/is/hail/expr/ir/FoldConstantsSuite.scala]: I removed this functionality from FoldConstants because it doesn't belong there (and Tim built it into a separate pass, ForwardLets).

File: hail/testng-cpp-codegen.xml (hunk 8 above)
Comment (cseed) [reply hail/testng-cpp-codegen.xml]: Oh, good catch!  I removed the tests from the pom.  I now set the flag in the cpp tests in the build.yaml.

**Comment by tpoterba** [comment]
Your errors are a result of -0.0, as https://github.com/hail-is/hail/issues/6086

**Comment by cseed** [comment]
I disabled the two tests that are failing because of this set/dict comparison issue.
